### PR TITLE
ARROW-13005: [C++] Add support for take implementation on dense union type

### PR DESCRIPTION
**Pull request by ZMZ91**
https://issues.apache.org/jira/browse/ARROW-13005

### CODE REVIEW

**Comment by ZMZ91** [comment]
ARROW-13005: [C++] Implementation of take function for dense union data type.

File: cpp/src/arrow/compute/kernels/vector_selection.cc [hunk 1]
```diff
@@ -1668,6 +1668,66 @@ struct ListImpl : public Selection<ListImpl<Type>, Type> {
...
+
+    DenseUnionArray typed_values(this->values);
+    auto num_fields = typed_values.num_fields();
+    ArrayVector child_arrays;
+    child_arrays.reserve(num_fields);
+    BufferVector buffers = {nullptr, checked_cast<const Int8Array&>(*child_ids).values(),
+                            checked_cast<const Int32Array&>(*value_offsets).values()};
+    *out = ArrayData(typed_values.type(), child_ids->length(), std::move(buffers), 0);
+    for (int i = 0; i < typed_values.num_fields(); i++) {
+      out->child_data.push_back(typed_values.field(i)->data());
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/vector_selection.cc:1725]: Hmm, if I understand correctly, you're reusing the full child arrays even if only a small fraction of them was taken? If so, this seems wasteful.

I think this can be done otherwise: maintain maps of the original offsets to the taken offsets, per child. This will admittedly use a lot of temporary storage (perhaps this can be done one child at a time?), but at least the final union array will be reasonably compact.

@bkietz What do you think?

File: cpp/src/arrow/compute/kernels/vector_selection.cc [hunk 2]
```diff
@@ -1668,6 +1668,66 @@ struct ListImpl : public Selection<ListImpl<Type>, Type> {
...
+    RETURN_NOT_OK(adapter.Generate(
+        [&](int64_t index) {
+          auto child_id = typed_values.child_id(index);
+          child_id_builder.UnsafeAppend(type_codes[child_id]);
+          auto value_offset = typed_values.value_offset(index);
+          value_offset_builder.UnsafeAppend(value_offset);
+          return Status::OK();
+        },
+        // TODO: not able to handle null case
+        VisitNoop));
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/vector_selection.cc:1701]: Well, either a null in the indices should raise `Status::NotImplemented`, or it should append a null in one of the child arrays (any of them, for example the first one by convention).

File: cpp/src/arrow/compute/kernels/vector_selection_test.cc [hunk 3]
```diff
@@ -1279,6 +1279,58 @@ TEST_F(TestTakeKernelWithStruct, TakeStruct) {
...
+    for (auto index_type : {int8(), uint32()}) {
+      ASSERT_OK(TakeJSON(type, values, index_type, indices, &actual));
+      ASSERT_OK(actual->Validate());
+      AssertArraysEqual(*expected, *actual, /*verbose=*/true);
+    }
+  }
+};
+
+TEST_F(TestTakeKernelWithDenseUnion, TakeDenseUnion) {
+  auto union_type = dense_union({field("a", int32()), field("b", utf8())}, {2, 5});
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/vector_selection_test.cc:1314]: Not sure why you're not reusing the test code from `DISABLED_TakeUnion` below? The example results seem different.

**Comment by pitrou** [comment]
@ZMZ91 Thanks for tackling this slightly delicate issue. I've posted two comments about the chosen approach.

File: cpp/src/arrow/type_traits.h [hunk 4]
```diff
@@ -435,6 +435,8 @@ struct TypeTraits<DenseUnionType> {
   using ArrayType = DenseUnionArray;
   using BuilderType = DenseUnionBuilder;
   using ScalarType = DenseUnionScalar;
+  using ValueOffsetBuilderType = Int32Builder;
+  using ChildIdBuilderType = Int8Builder;
```
Comment (bkietz) [inline cpp/src/arrow/type_traits.h:439]: These are not generally useful traits since they're not shared by other types. Please revert them

File: cpp/src/arrow/compute/kernels/vector_selection.cc (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/compute/kernels/vector_selection.cc:1725]: I think the larger problem here is that we may produce a dense union array whose offsets are not increasing, which is a violation of [the protocol](https://github.com/bkietz/arrow/blob/3c994458183a4585063a4925bad7bf02f29ab93c/docs/source/format/Columnar.rst#L561)

File: cpp/src/arrow/compute/kernels/vector_selection_test.cc [hunk 5]
```diff
@@ -1279,6 +1279,58 @@ TEST_F(TestTakeKernelWithStruct, TakeStruct) {
...
+        std::static_pointer_cast<DenseUnionArray>(ArrayFromJSON(type, values));
+    ArrayVector children;
+    children.reserve(type->num_fields());
+    std::vector<std::string> field_names;
+    field_names.reserve(type->num_fields());
+    for (int i = 0; i < type->num_fields(); i++) {
+      children.push_back(union_array->field(i));
+      field_names.push_back(type->field(i)->name());
+    }
+    ASSIGN_OR_ABORT(
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/vector_selection_test.cc:1297]:
```suggestion
    ASSERT_OK_AND_ASSIGN(
```

File: cpp/src/arrow/compute/kernels/vector_selection_test.cc [hunk 6]
```diff
@@ -1279,6 +1279,58 @@ TEST_F(TestTakeKernelWithStruct, TakeStruct) {
...
+        std::shared_ptr<Array> expected,
+        DenseUnionArray::Make(*ArrayFromJSON(int8(), type_codes),
+                              *ArrayFromJSON(int32(), value_offsets), children,
+                              field_names, union_array->union_type()->type_codes()));
+
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/vector_selection_test.cc:1302]:
```suggestion
    ASSERT_OK(expected->ValidateFull());
```

File: cpp/src/arrow/compute/kernels/vector_selection_test.cc [hunk 7]
```diff
@@ -1279,6 +1279,58 @@ TEST_F(TestTakeKernelWithStruct, TakeStruct) {
...
+    std::shared_ptr<Array> actual;
+
+    for (auto index_type : {int8(), uint32()}) {
+      ASSERT_OK(TakeJSON(type, values, index_type, indices, &actual));
+      ASSERT_OK(actual->Validate());
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/vector_selection_test.cc:1307]:
```suggestion
      ASSERT_OK(actual->ValidateFull());
```

**Review by bkietz — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/compute/kernels/vector_selection.cc (hunk 1 above)
Comment (ZMZ91) [reply cpp/src/arrow/compute/kernels/vector_selection.cc:1725]: Really appreciate for both your reviews. This is a quick implementation of take function on dense union data type. But as you see there are some problems.

@pitrou yes, this implementation will copy all child arrays and as you mentioned below, it's hard to handle null case. Because it doesn't maintain child builders and there may be no nulls in the first or any child array.

@bkietz if the indices from the argument not in asc order or the original dense union array not following the protocol, the value offsets for the generated array may not increase.

And I've thought about @pitrou 's solution that maintaining all child builders there. I agree that's a better solution but not very clear how to handle all different kinds of child builders. They could be any type, including complex ones like map, list or event union it self. And the implementation here sounds like a big switch case to me. Maybe some macro helps. Is there some example in existing code that handles various array builders could be referred to here?

Comment (pitrou) [reply cpp/src/arrow/compute/kernels/vector_selection.cc:1725]: I don't think you need to use any child builders. Instead, you should be able to call `Take` on each of the child arrays; you just have to construct the indices carefully for each child array.

Comment (ZMZ91) [reply cpp/src/arrow/compute/kernels/vector_selection.cc:1725]: But for the null case, it still needs to a null in the first child, right? And there may be no nulls in the first child.

Comment (pitrou) [reply cpp/src/arrow/compute/kernels/vector_selection.cc:1725]: You can put a null in the indices.

**Comment by ZMZ91** [comment]
Hi @pitrou and @bkietz, just update the pr due to your comments. Please help check the latest code. Thanks a lot.

File: cpp/src/arrow/compute/kernels/vector_selection.cc [hunk 8]
```diff
@@ -1668,6 +1668,82 @@ struct ListImpl : public Selection<ListImpl<Type>, Type> {
...
+    RETURN_NOT_OK(value_offset_builder.Finish(&value_offsets));
+
+    DenseUnionArray typed_values(this->values);
+    auto num_fields = typed_values.num_fields();
+    BufferVector buffers = {nullptr, checked_cast<const Int8Array&>(*child_ids).values(),
+                            checked_cast<const Int32Array&>(*value_offsets).values()};
+    *out = ArrayData(typed_values.type(), child_ids->length(), std::move(buffers), 0);
+    for (auto i = 0; i < num_fields; i++) {
+      std::shared_ptr<Int32Array> child_indices_array;
+      RETURN_NOT_OK(child_indices_builders[i]->Finish(&child_indices_array));
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/vector_selection.cc:1738]: Nit: prefer result returning overloads
```suggestion
      ARROW_ASSIGN_OR_RAISE(auto child_indices_array,
                            child_indices_builders[i]->Finish());
```

File: cpp/src/arrow/compute/kernels/vector_selection.cc [hunk 9]
```diff
@@ -1668,6 +1668,82 @@ struct ListImpl : public Selection<ListImpl<Type>, Type> {
   }
 };
 
+struct DenseUnionImpl : public Selection<DenseUnionImpl, DenseUnionType> {
+  using Base = Selection<DenseUnionImpl, DenseUnionType>;
+  LIFT_BASE_MEMBERS();
+
+  Int32Builder value_offset_builder;
+  Int8Builder child_id_builder;
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/vector_selection.cc:1676]: It's worth noting that ArrayBuilder always produces a null bitmap, which is not necessary for the buffers these are used to produce. See also: TypedBufferBuilder

File: cpp/src/arrow/compute/kernels/vector_selection_test.cc [hunk 10]
```diff
@@ -1281,33 +1280,32 @@ TEST_F(TestTakeKernelWithStruct, TakeStruct) {
...
+// TODO: Restore Sparse Union take functionality
+TEST_F(TestTakeKernelWithUnion, TakeUnion) {
+  auto union_type = dense_union({field("a", int32()), field("b", utf8())}, {2, 5});
+  auto union_json = R"([
       null,
       [2, 222],
       [5, "hello"],
       [5, "eh"],
       null,
       [2, 111]
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/vector_selection_test.cc:1292]: Since union doesn't support top level nulls, please be explicit about the type_id associated with a null and test with nulls in each child array
```suggestion
      [2, null],
      [2, 222],
      [5, "hello"],
      [5, "eh"],
      [2, null],
      [2, 111],
      [5, null]
```

**Review by bkietz — CHANGES_REQUESTED** [review]

**Comment by ZMZ91** [comment]
Thank you @bkietz. All comments resolved. Please help check.

File: cpp/src/arrow/compute/kernels/vector_selection.cc [hunk 11]
```diff
@@ -1668,6 +1668,80 @@ struct ListImpl : public Selection<ListImpl<Type>, Type> {
...
+    return Status::OK();
+  }
+
+  Status Finish() override {
+    ARROW_ASSIGN_OR_RAISE(auto child_ids_buffer, child_id_buffer_builder_.Finish());
+    ARROW_ASSIGN_OR_RAISE(auto value_offsets_buffer,
+                          value_offset_buffer_builder_.Finish());
+    DenseUnionArray typed_values(this->values);
+    auto num_fields = typed_values.num_fields();
+    BufferVector buffers = {nullptr, child_ids_buffer, value_offsets_buffer};
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/vector_selection.cc:1731]:
```suggestion
    BufferVector buffers{nullptr, std::move(child_ids_buffer), std::move(value_offsets_buffer)};
```

File: cpp/src/arrow/compute/kernels/vector_selection_test.cc [hunk 12]
```diff
@@ -607,31 +607,29 @@ TEST_F(TestFilterKernelWithStruct, FilterStruct) {
...
+TEST_F(TestFilterKernelWithUnion, FilterUnion) {
+  auto union_type = dense_union({field("a", int32()), field("b", utf8())}, {2, 5});
+  auto union_json = R"([
+      [2, null],
       [2, 222],
       [5, "hello"],
       [5, "eh"],
-      null,
+      [2, null],
       [2, 111]
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/vector_selection_test.cc:618]: To test that nulls in other than the first field work, please add:
```suggestion
      [2, 111],
      [5, null]
```

**Comment by ZMZ91** [comment]
Hi @bkietz and @pitrou, just a question not quite related to this pr and about dense union array. I've seen the limitation for array data is under 2GB. Does it apply to nested arrays or just the sub arrays inside? For example, dense union array, it has child id array, value offset array and a bunch of child arrays. Each of them should be under 2GB. And is it restricted that the sum under 2GB?

**Comment by bkietz** [comment]
The 2GB limitation applies to buffers and is type specific. It derives from usage of signed 32 bit integers to encode offsets, so for example a StringArray cannot have more than 2GB total character data because otherwise we could not express the final offset. IIUC it's allowed for a child of the union to be an Int8Array with `2**33` elements but a dense union's offsets could not point to most of those elements.

In short, this is not a constraint you need to enforce in this PR.

File: cpp/src/arrow/compute/kernels/vector_selection.cc [hunk 13]
```diff
@@ -1668,6 +1668,81 @@ struct ListImpl : public Selection<ListImpl<Type>, Type> {
...
+  Status GenerateOutput() {
+    DenseUnionArray typed_values(this->values);
+    Adapter adapter(this);
+    RETURN_NOT_OK(adapter.Generate(
+        [&](int64_t index) {
+          int8_t child_id = typed_values.child_id(index);
+          child_id_buffer_builder_.UnsafeAppend(type_codes_[child_id]);
+          int32_t value_offset = typed_values.value_offset(index);
+          value_offset_buffer_builder_.UnsafeAppend(
+              child_indices_builders_[child_id].length());
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/vector_selection.cc:1702]:
```suggestion
              static_cast<int32_t>(child_indices_builders_[child_id].length()));
```
This is implicit conversion from `int64_t -> int32_t`. Unfortunately it needs to be explicit: https://ci.appveyor.com/project/ApacheSoftwareFoundation/arrow/builds/39924781/job/kkm24qqycxd29y5f#L1096

File: cpp/src/arrow/compute/kernels/vector_selection.cc [hunk 14]
```diff
@@ -1668,6 +1668,81 @@ struct ListImpl : public Selection<ListImpl<Type>, Type> {
...
+          RETURN_NOT_OK(child_indices_builders_[child_id].Reserve(1));
+          child_indices_builders_[child_id].UnsafeAppend(value_offset);
+          return Status::OK();
+        },
+        [&]() {
+          int8_t child_id = 0;
+          child_id_buffer_builder_.UnsafeAppend(type_codes_[child_id]);
+          value_offset_buffer_builder_.UnsafeAppend(
+              child_indices_builders_[child_id].length());
```
Comment (bkietz) [inline cpp/src/arrow/compute/kernels/vector_selection.cc:1711]:
```suggestion
          value_offset_buffer_builder_.UnsafeAppend(
              static_cast<int32_t>(child_indices_builders_[child_id].length()));
```

**Review by bkietz — CHANGES_REQUESTED** [review]

**Comment by ZMZ91** [comment]
Hi @bkietz and @pitrou, could you help check what else to update in this pr? Or we may merge it? Thanks.

**Review by bkietz — APPROVED** [review]
LGTM, thanks!

I'll merge on green CI
